### ISSUE TITLE
Heating and Cooling Thresholds always read 80°; Illegal Heat Threshold Values in Log

### ISSUE THREAD

**Issue by johnataravir** [dotfortun3-code/homebridge-infinitude-v2#9]
Hi! I've had some success with the plugin: I can use it to shut my Carrier off, and change it between heating and cooling. It reads back the temperature and mode correctly to HomeKit. But I can't seem to set the heating or cooling set points. No matter what the Carrier reads, HomeKit always reads the set point at 80° F.

Possibly related, I get these log entries frequently. I turned on debugging to get the stack trace:

```
[8/29/2022, 9:11:44 PM] [homebridge-infinitude-v2] This plugin generated a warning from the characteristic 'Heating Threshold Temperature': characteristic was supplied illegal value: number 26.666666666666664 exceeded maximum of 25. See https://homebridge.io/w/JtMGR for more info.
[8/29/2022, 9:11:44 PM] [homebridge-infinitude-v2] Error: 
    at HeatingThresholdTemperature.Characteristic.characteristicWarning (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:2142:105)
    at HeatingThresholdTemperature.Characteristic.validateUserInput (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:2049:14)
    at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:1617:24
    at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/util/once.ts:10:18
    at InfinitudeThermostat.<anonymous> (/homebridge/node_modules/homebridge-infinitude-v2/src/InfinitudeThermostat.js:168:15)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

26.66° C == 80° F, so the error is probably connected to the display issue.

That log was from just a few minutes ago, but the Carrier is in cooling mode. So I'd expect a warning about the cooling threshold, not the heating threshold.

**Comment by dotfortun3-code** [comment]
@johnataravir Is the thermostat in Celsius or Fahrenheit? Also, can you check what Infinitude is reporting as your heat setpoint? My current thought is it might be a conversion issue, but this information will help.

**Comment by johnataravir** [comment]
The thermostat is in Fahrenheit. Both the heat and cooling setpoints are reported correctly in Infinitude (71 for cool at the time I reported, and 54 for heat). I confirmed that in the API, too- and I double-checked that the JSON path I’m getting back from the API corresponds to what the plugin is reading. It looks to me like it is.

**Comment by johnataravir** [comment]
Oh- also, I made a mistake above. I confirmed last night that the plugin is setting the setpoints correctly. The only problem is that once it does, HomeKit still reads them as set to 80.

**Comment by dotfortun3-code** [comment]
Could you drop your status.json from the infinitude instance? You can get it by going to http://infinitude-ip:port/status.json.

**Comment by johnataravir** [comment]
Sure thing! It's enclosed. Github "doesn't support that format" (JSON), so I changed the extension to .txt.

[status.txt](https://github.com/dotfortun3-code/homebridge-infinitude-v2/files/9456098/status.txt)

**Comment by johnataravir** [comment]
So, this is odd. I grabbed and uploaded the above status.json a minute ago, and uploaded without looking at it. Then I went upstairs, and realized it was really hot. Sure enough, the Carrier was actually set to a CLSP of 80°, and the output agrees. So I'm enclosing the one I took last night, as well. I verified that HomeKit said the CLSP was 80° when I took this one, but you can see that the CLSP for Zone 1 is 71°.
[status-original.txt](https://github.com/dotfortun3-code/homebridge-infinitude-v2/files/9456238/status-original.txt)

I'm not sure what to make of that. Maybe HomeKit is somehow getting 80° as the CLSP, and is then setting it again when it sees drift?

I'm wary of changing too many variables at once, but I haven't blocked the API servers, so the pump is still talking to Carrier. I know Infinitude should bypass all that, but let me know if you think I should block those. It might eliminate a potential variable.

**Comment by dotfortun3-code** [comment]
I found something interesting. The JSON you sent me shows a CLSP and HTSP of 80°, but I am unable to actually get Infinitude to take 80° for the heat set point. I am not sure what is going on there.

I do have an update that might fix the issue. The 80° heat set point might be soft locking the plugin, causing it to not read and write other values correctly. The new version should be on NPM shortly. Let me know if it fixes your issue. It fixes some other small issues.

**Comment by johnataravir** [comment]
Awesome! Thanks so much for the fast response on this. I'll test it out as soon as I can and let you know how it goes.

**Comment by dotfortun3-code** [comment]
No problem. If this patch doesn't work, we can try blocking the Carrier API, but I have not had to do that for the plugin to work, but it might comedown to variances in hardware/system configuration.

**Comment by johnataravir** [comment]
I'm running the patched version now. It seems like HomeKit is now unable to change the set point at all. It can still turn the device on and off, but no matter what I do, the set point seems fixed at 80 (which appears to be some kind of default in the Carrier: it's set at 80 whenever I turn it back on from being off).

I'm also seeing this error in the logs, which seems new:

```
[8/30/2022, 9:07:34 PM] [homebridge-infinitude-v2] This plugin generated a warning from the characteristic 'Target Temperature': characteristic value expected valid finite number and received "[object Promise]" (object). See https://homebridge.io/w/JtMGR for more info.
[8/30/2022, 9:07:34 PM] [homebridge-infinitude-v2] Error: 
    at TargetTemperature.Characteristic.characteristicWarning (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:2142:105)
    at TargetTemperature.Characteristic.validateUserInput (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:2030:14)
    at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:1617:24
    at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/util/once.ts:10:18
    at InfinitudeThermostat.<anonymous> (/homebridge/node_modules/homebridge-infinitude-v2/src/InfinitudeThermostat.js:59:17)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

**Comment by dotfortun3-code** [comment]
I took some time to review the JSON you sent in more detail, and there are some differences between it and what I see on my local system. One thing that stands out to me is you version is 1.41 and mine is 1.7. I am not exactly sure where Infinitude pulls this, but what thermostat do you have (model #) and what version/firmware is it on? I believe you can check in one of the setting menus on the t-stat itself. It might be a compatibility issue with either newer/older hardware.

Also, if you set the heating/cooling set point directly with Infinitude, does it set it correctly? I want to try to reduce the number of possible breakpoints and rule out infinitude as the culprit.

One final thing, do you know how to install plugins manually to Homebridge? I can create a branch you can grab to test changes from, instead of dropping updates on the main plugin.

**Comment by johnataravir** [comment]
Interesting. I've confirmed that the system thinks it's up to date, at least. It did an OTA update within the last month.

![IMG_4328](https://user-images.githubusercontent.com/3493026/187799932-4550d7e0-4865-41b2-9b3b-28119b108b91.jpeg)

Probably irrelevant, this is the information on the heat pump:
model: 25VNA024A00300
software version: CESR131438-09

And this is what it shows for the Infinity Touch controller, which is a "Series B":
model: SYSTXCCITC01-B
software version: 131626-04.05

I've confirmed that setting the set points directly with Infinitude works fine.

I came across [these instructions](https://github.com/homebridge/homebridge) on using npm to run a non-prod version of the plugin, but couldn't get them to work with the auto-updated Docker setup I have. If you have some instructions or anything you think would help, I'd happily test from a new branch.

Thanks for your help with this!

**Comment by dotfortun3-code** [comment]
This is good information. You definitely have a different version of the t-stat than mine.

Could you provide the json you get  from http://<infinitude-ip>:<port>/api/config and http://<infinitude-ip>:<port>/api/status?

I noticed the JSON from the status.json endpoint can be different from those endpoints and those are what the plugin uses.

Since infinitude can set the temperatures correctly, my guess is the data is different than what the plug-in is expecting. Hopefully, the config/status JSON will provide some more insight.

As for the manual plugin setup, it can be pretty complicated (especially with docker, since symlinks are weird). I can provide instructions if there isn't something clear I can get from the config and json.

**Comment by johnataravir** [comment]
Makes sense, and sounds good re: the manual plugin stuff. Here are the files:

[api-status.txt](https://github.com/dotfortun3-code/homebridge-infinitude-v2/files/9473477/api-status.txt)
[api-config.txt](https://github.com/dotfortun3-code/homebridge-infinitude-v2/files/9473478/api-config.txt)

**Comment by dotfortun3-code** [comment]
I wasn't able to find anything explicit in those. I can say there are some additional parameters in you config/status calls, but I am not sure that it means that is the issue. I released a new version that adds a new option in the plugin config to enable verbose logging. If you could turn that on and try to set the temperature from HK and send the log file that would be helpful. You can find the log in the root of the homebridge install (if using docker, it's where you mounted the Docker volume). The file should be called homebridge.log.

**Comment by johnataravir** [comment]
Thanks! I updated the plugin, restarted, and to get some good logs, I tried to set the cooling threshold to 63°— and it worked. So I turned it off, then set it to cool to 67°, which is the value I'd tested so many times before, and that worked, too. I didn't change anything else.

Unfortunately, it looks like homebridge.log contains all the logs from my homebridge system, some of which are more sensitive than I'd want to post on Github. Is there an address I could send them to you at?

I did capture this, which is what the plugin printed just after I set the heat to 63° for the test. I noticed that the clsp isn't exactly 63, and it doesn't correspond to a clean number in Celsius, either. (It's 17.2° C.)

```
[9/5/2022, 7:10:04 PM] [homebridge-infinitude-v2] [VERBOSE] getTemperatureScale() response: F
[9/5/2022, 7:10:04 PM] [homebridge-infinitude-v2] [VERBOSE] http://unicron:3000/api/1/activity/home?clsp=68.00 response: {"clsp":["68.00"],"fan":["high"],"htsp":["55.0"],"id":"home"}
[9/5/2022, 7:10:40 PM] [homebridge-infinitude-v2] [VERBOSE] getTemperatureScale() response: F
[9/5/2022, 7:10:40 PM] [homebridge-infinitude-v2] [VERBOSE] http://unicron:3000/api/1/activity/home?clsp=62.96 response: {"clsp":["62.96"],"fan":["high"],"htsp":["55.0"],"id":"home"}
[9/5/2022, 7:12:27 PM] [homebridge-infinitude-v2] [VERBOSE] getTemperatureScale() response: F
[9/5/2022, 7:12:27 PM] [homebridge-infinitude-v2] [VERBOSE] getTemperatureScale() response: F
[9/5/2022, 7:12:27 PM] [homebridge-infinitude-v2] [VERBOSE] getTemperatureScale() response: F
[9/5/2022, 7:12:27 PM] [homebridge-infinitude-v2] [VERBOSE] getTemperatureScale() response: F
[9/5/2022, 7:12:27 PM] [homebridge-infinitude-v2] [VERBOSE] GetTemperatures(): Found zone with ID 1: {"clsp":["63.0"],"currentActivity":["home"],"damperposition":["15"],"enabled":["on"],"fan":["high"],"hold":["off"],"htsp":["55.0"],"id":"1","name":["ZONE 1"],"occupancy":["unoccupied"],"occupancyOverride":["off"],"otmr":[{}],"rh":["67"],"rt":["69.0"],"zoneconditioning":["idle"]}
[9/5/2022, 7:12:27 PM] [homebridge-infinitude-v2] [VERBOSE] GetTemperatures(): Response to HomeKit {"htsp":12.78,"clsp":17.22,"currentTemp":{},"mode":"cool"}
[9/5/2022, 7:12:27 PM] [homebridge-infinitude-v2] [VERBOSE] GetTemperatures(): Found zone with ID 1: {"clsp":["63.0"],"currentActivity":["home"],"damperposition":["15"],"enabled":["on"],"fan":["high"],"hold":["off"],"htsp":["55.0"],"id":"1","name":["ZONE 1"],"occupancy":["unoccupied"],"occupancyOverride":["off"],"otmr":[{}],"rh":["67"],"rt":["69.0"],"zoneconditioning":["idle"]}
[9/5/2022, 7:12:27 PM] [homebridge-infinitude-v2] [VERBOSE] GetTemperatures(): Response to HomeKit {"htsp":12.78,"clsp":17.22,"currentTemp":{},"mode":"cool"}
[9/5/2022, 7:12:27 PM] [homebridge-infinitude-v2] [VERBOSE] GetTemperatures(): Found zone with ID 1: {"clsp":["63.0"],"currentActivity":["home"],"damperposition":["15"],"enabled":["on"],"fan":["high"],"hold":["off"],"htsp":["55.0"],"id":"1","name":["ZONE 1"],"occupancy":["unoccupied"],"occupancyOverride":["off"],"otmr":[{}],"rh":["67"],"rt":["69.0"],"zoneconditioning":["idle"]}
[9/5/2022, 7:12:27 PM] [homebridge-infinitude-v2] [VERBOSE] GetTemperatures(): Response to HomeKit {"htsp":12.78,"clsp":17.22,"currentTemp":{},"mode":"cool"}
[9/5/2022, 7:12:27 PM] [homebridge-infinitude-v2] [VERBOSE] getTemperatureScale() response: F
```

Let me know what you think! If I turn verbose logging back off and it still works, then my problem is solved, technically, anyway. I can always keep an eye on it and get back in touch if it stops working again.

(Enabling logging masked a bug in something I worked on, once. Printing set an uninitialized value to empty, or something- it was a long time ago. I don't think that's happening here, but since then, I've always been careful about setting my config back to a production state and testing one last time.)

**Comment by dotfortun3-code** [comment]
One thing I also changed with this patch that might have been the issue is rounding to two decimal places, so it's possible that the version of the thermostat you have doesn't like going beyond a certain number of decimal places. There was a bug where it wasn't rounding all the time. I will probably bump that down to a single decimal place, I just chose two out of habit haha.

I shouldn't need the homebridge.log

I will close this for now. If you disable logs and it breaks, let me know.
